### ISSUE TITLE
WCAG2.2 SC 2.5.2 violation - Dropdown Menu opens on `onPointerDown`

### ISSUE THREAD

**Issue by bolonio** [radix-ui/primitives#3124]
## Bug report

The [Radix Dropdown Menu component](https://www.radix-ui.com/primitives/docs/components/dropdown-menu) opens on [`onPointerDown`](https://github.com/radix-ui/primitives/blob/main/packages/react/dropdown-menu/src/DropdownMenu.tsx#L116) (which means as soon as a mouse navigation user presses the triggering element of the dropdown menu), but there is not available a function on the [up-event](https://www.w3.org/WAI/WCAG22/Understanding/pointer-cancellation.html#dfn-up-event), or a [mechanism](https://www.w3.org/WAI/WCAG22/Understanding/pointer-cancellation.html#dfn-mechanism) to abort the function before completion or to undo the function after completion. You can take a look at the [Understanding SC 2.5.2: Pointer Cancellation (Level A)](https://www.w3.org/WAI/WCAG22/Understanding/pointer-cancellation.html) article for more information.

> The intent of this success criterion is to make it easier for users to prevent accidental or erroneous pointer input. People with various disabilities can inadvertently initiate touch or mouse events with unwanted results. Each of the following subsections roughly aligns with the bullets of this Success Criterion, and outlines a means of allowing users to cancel pointer operations.
>
> Source: [Understanding SC 2.5.2: Pointer Cancellation (Level A)
](https://www.w3.org/WAI/WCAG22/Understanding/pointer-cancellation.html)

### Current behavior

1. A mouse navigation user clicks (and holds) the triggering element of the dropdown menu.
2. Without releasing, moves the mouse outside of both the triggering element or the dropdown menu area.
3. The dropdown menu is not closed (cancelling the function).

![CleanShot 2024-09-18 at 15 40 57](https://github.com/user-attachments/assets/1fa5c094-7684-4812-b73c-c2645e215c93)

### Expected behavior

1. A mouse navigation user clicks (and holds) the triggering element of the dropdown menu.
2. The dropdown menu is not yet open.
3. The user releases the mouse click while still hovering the triggering element of the dropdown menu.
4. The dropdown menu opens.

![CleanShot 2024-09-18 at 15 45 29](https://github.com/user-attachments/assets/c7ddced7-5676-42aa-b12f-b2855789f377)

You can see the behaviour in the [W3C Menu Button Pattern](https://www.w3.org/WAI/ARIA/apg/patterns/menu-button/) or in the [W3C Menu Bar Pattern](https://www.w3.org/WAI/ARIA/apg/patterns/menubar/)

![CleanShot 2024-09-18 at 15 42 00](https://github.com/user-attachments/assets/1e2ad108-aed2-445f-b10b-cee3f90ffba2)

or moving the mouse outside of the dropdown menu triggering and content should cancel the `onPointerDown` function.

1. A mouse navigation user clicks (and holds) the triggering element of the dropdown menu.
2. The dropdown menu opens.
3. Without releasing, moves the mouse outside of both the triggering element or the dropdown menu area.
4. The dropdown menu closes.
5. The user releases the mouse click while hovering outside of both the triggering element or the dropdown menu area.
6. The dropdown menu is still closed.

### Reproducible example

[Radix Dropdown Menu component](https://www.radix-ui.com/primitives/docs/components/dropdown-menu) 

### Suggested solution

Change the opening of the dropdown menu to `onPointerUp` or `onClick` instead of using `onPointerDown`.

### Additional context

It might be related to https://github.com/radix-ui/primitives/issues/2700 (not 100% sure about it though)

### Your environment

| Software         | Name(s) | Version |
| ---------------- | ------- | ------- |
| Radix Package(s) | @radix-ui/react-dropdown-menu | ^2.0.5|
| React            | n/a     | 18.2.0 |
| Browser          | Google Chrome | 128.0.6613.138 |
| Assistive tech   |         |         |
| Node             | n/a     |         |
| npm/yarn         |         |         |
| Operating System | MacOS | 14.6.1 |

**Comment by bolonio** [comment]
There's a [possible "workaround"](https://github.com/radix-ui/primitives/issues/2418) that might work (from the side of the consumer), but unfortunately it will only work for mouse users, it breaks the keyboard navigation. I'm still investigating if there's a nicer workaround for both cases.

**Comment by wintercounter** [comment]
I have a solution that's going to stay backward compatible and won't break anything.

```
const handler = useRef("pointer")

onPointerDown={(ev) => {
   if (ev.pointerType === "touch") {
       handler.current = "click"
       document.body.addEventListener("pointerup", () => {
           queueMicrotask(() => {
               handler.current = "pointer"
           })
       }, { once: true })
       return
   }
   // ...old logic
}}
onClick={() => {
   if (handler.current !== 'click') return
   // ...old logic
}}
```

So basically if `pointerdown` is initiated by touch, listen to the click handler instead.

This is the event order, which is the reason queueMicrotask is necessary:

```
pointerdown
pointerup
click
```

Happy to make a PR if this solution is acceptable.

**Comment by programming-with-ia** [comment]
+1

**Comment by NoraHNIKT** [comment]
+1

**Comment by asumaran** [comment]
I’d go even further and suggest exposing a way to choose which event should trigger the dropdown. In my case, I had a button that needed to be draggable with dnd-kit and also show the Radix dropdown on right-click. There was a conflict because both libraries use the onPointerDown event, so attempting to drag the button would open the dropdown, with no apparent way to change this behavior.

I propose there should be an option on the `DropdownMenu.Trigger` component to define which event should be used to trigger the dropdown.
